sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1774,6 +1775,34 @@ class Analyzer(
   }
 
   /**
+   * Substitute Hints.
+   */
+  object SubstituteHint extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan transform {
```

Comment:
Nested one works like the design. I'm making testcase to ensure that.
Here is a example for your query.
```scala
spark.range(10).createOrReplaceTempView("tbl_a")
spark.range(20).createOrReplaceTempView("tbl_b")
spark.range(30).createOrReplaceTempView("tbl_c")

val plan = sql(
      """SELECT /*+ MAPJOIN(tbl_a, tbl_a) */
        |       *
        |FROM   tbl_a A
        |       JOIN tbl_b B
        |        ON B.id = A.id
        |       JOIN (SELECT /*+ MAPJOIN(tbl_c) */
        |                         XA.id
        |                  FROM   tbl_b XA
        |                          LEFT SEMI JOIN tbl_c XB
        |                          ON XB.id = XA.id) C
        |        ON C.id = A.id
      """.stripMargin).queryExecution.optimizedPlan

'Join Inner, (id#0L = id#0L)
:- 'Join Inner, (id#0L = id#0L)
:  :- BroadcastHint
:  :  +- Range (0, 10, splits=2)
:  +- Range (0, 20, splits=2)
+- 'Join LeftSemi, (id#0L = id#0L)   
   :- Range (0, 20, splits=2)        
   +- BroadcastHint                  
      +- Range (0, 30, splits=2)   
```